client/src/redux/actions/credentialsActions.js

```diff
@@ -33,9 +33,51 @@ const getCredentials = () => {
   };
 };
 
+const addCredentialLoading = bool => ({
+  type: credentialsTypes.ADD_CREDENTIAL_LOADING,
+  loading: bool
+});
+
+const addCredentialSuccess = data => ({
+  type: credentialsTypes.ADD_CREDENTIAL_SUCCESS,
+  data
+});
+
+const addCredentialError = (bool, message) => ({
+  type: credentialsTypes.ADD_CREDENTIAL_ERROR,
+  error: bool,
+  message: message
+});
+
+const addCredential = (data, addCallback) => {
+  return function(dispatch) {
+    dispatch(addCredentialLoading(true));
+    return credentialsService
+      .addCredential(data)
+      .then(success => {
+        dispatch(addCredentialSuccess(success));
+        dispatch(addCredentialLoading(false));
+        if (addCallback) {
+          addCallback(false, success);
+        }
+      })
+      .catch(error => {
+        dispatch(addCredentialError(true, error.message));
+        dispatch(addCredentialLoading(false));
+        if (addCallback) {
+          addCallback(true, error.message);
+        }
+      });
+  };
+};
+
```

Comment:
Note: we need to re-evaluate the callback setup... I'm thinking next round either we return a promise instead, that way we can remove the checks/logic around it... Or we figure out how to work with state value observation

Reply:
not sure why I didn't recommend this at first, but @jeff-phillips-18 this is probably the use case for using [Redux ```Store.subscribe```](https://redux.js.org/docs/api/Store.html#subscribe)